src/checkout.c

```diff
@@ -1794,6 +1794,9 @@ static int checkout_create_the_new(
 	int error = 0;
 	git_diff_delta *delta;
 	size_t i;
+	int caps = git_index_caps(data->index);
+
+	git_index_set_caps(data->index, caps & !GIT_INDEXCAP_NO_FILEMODE);
```

Comment:
This looks very suspicious. `!GIT_INDEXCAP_NO_FILEMODE` is the logical negation of this flag. As its value is 2, this would make this expression `caps & 0` thus negating all of the caps. You're turning off `NO_FILEMODE` but that's just an accident. What you want to write is likely `caps & (~GIT_INDEXCAP_NO_FILEMODE)` which will pass through any bits set in `caps` except for the `NO_FILEMODE` one.